### PR TITLE
Change platforms to multiple if duplicate tickets have a different platform (MEQS_KEEP_PLATFORM to prevent changes)

### LINKED ISSUE #430: Add MEQS_KEEP_PLATFORM to help prevent incorrect platform change / generalization
## The Problem
It is not always clear whether a bug is intended to be platform-specific.

## Possible Solution
If MEQS_KEEP_PLATFORM is set:

1, If someone changes the platform directly (usually to "Multiple") Arisa would either send a message to the helper/mod who did that, or revert the change and add a comment if was by a user.
2. If a duplicate is linked that has a different platform, the moderator making the link is notified.

Depends on Issue #421 being implemented.

### LINKED ISSUE #431: Automatically change platform to "multiple" if linked duplicate has different platform
## The Problem
Most bugs are not platform-specific, but the reporter should usually choose their own platform. It takes time to then look at the duplicates to see if they reference a different plaform

## Possible Solution
Automatically change platform to "multiple" unless MEQS_KEEP_PLATFORM is set (Issue #430) and a duplicate is linked with a different platform

### PR DESCRIPTION
**Pull request by chandler05**
## Purpose
Closes #431 and closes #430
## Approach
Copied and changed the ConfirmParentModule, and the KeepPrivateModule
Added in the Platform field

### CODE REVIEW

**Comment by chandler05** [comment]
PR is ready for review 👍

**Review by urielsalis — APPROVED** [review]

File: src/main/kotlin/io/github/mojira/arisa/modules/MultiplePlatformsModule.kt [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+    }
+
+    private fun isDuplicatedWithDifferentPlatforms(platform: String?, blacklist: List<String>, issue: Issue):
+            Either<ModuleError, Boolean> = Either.fx {
+        issue.links
+            .filter(::isDuplicatedLink)
+            .forEach {
+                val child = it.issue.getFullIssue().toFailedModuleEither().bind()
+                if (child.platform !in blacklist && child.platform != platform.getOrDefault("None")) {
+                    return@fx true
```
Comment (urielsalis) [inline src/main/kotlin/io/github/mojira/arisa/modules/MultiplePlatformsModule.kt:34]: I don't think you need a FX block here, its really heavy on the call stack(3 levels causes a stackoverflow)

Comment (chandler05) [reply src/main/kotlin/io/github/mojira/arisa/modules/MultiplePlatformsModule.kt:34]: That's how it's written in the ConfirrmParentModule

Comment (chandler05) [reply src/main/kotlin/io/github/mojira/arisa/modules/MultiplePlatformsModule.kt:34]: That's where I took it from

Comment (urielsalis) [reply src/main/kotlin/io/github/mojira/arisa/modules/MultiplePlatformsModule.kt:34]: maybe both shouldnt have it :P

Comment (urielsalis) [reply src/main/kotlin/io/github/mojira/arisa/modules/MultiplePlatformsModule.kt:34]: the only reason is there is                 val child = it.issue.getFullIssue().toFailedModuleEither().bind()

You can just do an if for that (without toFailedModuleEither)

**Comment by MikeUnverferth** [comment]
I think this should not be merged until issue #421 is implemented. Otherwise we won't know that there are unwanted platform conflicts on an issue with KEEP_PLATFORM set.

**Comment by chandler05** [comment]
#421 is implemented is this PR already

**Comment by chandler05** [comment]
That can most likely be handled separately. A lot of the modules already have similar consequences, and we've never had a command for direct messages. It just seems like it's not necessary in order to implement this PR, but it would be helpful in the future.

File: src/main/kotlin/io/github/mojira/arisa/modules/MultiplePlatformsModule.kt (hunk 1 above)
Comment (chandler05) [reply src/main/kotlin/io/github/mojira/arisa/modules/MultiplePlatformsModule.kt]: without doesn't exist? Plus, for some reason the platform and reporter fields are recognized by the code unless .bind() is attached to the end of the val child line

Comment (chandler05) [reply src/main/kotlin/io/github/mojira/arisa/modules/MultiplePlatformsModule.kt:33]: I'm going to leave as-is right now unless its very necessary to do this

**Comment by MikeUnverferth** [comment]
If there's already precedent for doing it that way then it's ok with me.

**Comment by MikeUnverferth** [comment]
Could a message be added to arisa-error-log when it does not change the platform due to KEEP_PLATFORM being set? Though it's not technically an error, it's not a normal result either and it might get lost in the routine arisa-log.

**Comment by urielsalis** [comment]
No. arisa-error-log is for errors we should act on(usually immideately), not for expected cases

File: src/main/kotlin/io/github/mojira/arisa/modules/MultiplePlatformsModule.kt (hunk 1 above)
Comment (urielsalis) [reply src/main/kotlin/io/github/mojira/arisa/modules/MultiplePlatformsModule.kt:34]: when I said without I mean removing .toFailedModuleEither() and just doing the if yourself

By the way, it doesnt really make sense to convert it to a either anyways

**Comment by urielsalis** [comment]
And I do think we should add a warning message, but only if a volunteer/reporter tries to change it with the tag set(not on the duplicate thing). Same warning as in future version

File: arisa.json [hunk 2]
```diff
@@ -175,6 +176,27 @@
...
+          "Windows 10 Mobile",
+          "Xbox One"
+        ],
+        "targetPlatform": ["Multiple"],
+        "transferredPlatformBlacklist": ["None"],
+        "keepPlatformTag": ["MEQS_KEEP_PLATFORM"],
+        "whitelist": ["MCPE"]
+      },
+      "keepPlatform": {
+        "keepPlatformTag": ["MEQS_KEEP_PLATFORM"],
```
Comment (urielsalis) [inline arisa.json:197]: Why is this a list? 🤔

**Comment by chandler05** [comment]
I could never figure out how to add messages in, so that's why it doesn't have one

**Comment by MikeUnverferth** [comment]
Can it write a [WARNING] entry to arisa-log? Same would be for whatever other "consequences" that other modules might do that a mod should be made aware of.

**Comment by urielsalis** [comment]
A log is fine(info), but I really dont think its necessary

File: arisa.json (hunk 2 above)
Comment (chandler05) [reply arisa.json:197]: Which part shouldn't be a list?

Comment (urielsalis) [reply arisa.json:197]: keepPlatformTag

Comment (chandler05) [reply arisa.json:197]: Is it? TargetPlatform is the same...
I... Uh, got it from the Arisa.json for ConfirmParent and KeepPrivate, and they're not lists but are written that way

Comment (urielsalis) [reply arisa.json:197]: keepPlatformTag": "MEQS_KEEP_PLATFORM" to make sure no one thinks we can put more than 1 (and read as String, not List<String>)

Comment (chandler05) [reply arisa.json:197]: Should I remove it from the target platform as well?

Comment (urielsalis) [reply arisa.json:197]: yep

**Comment by MikeUnverferth** [comment]
If a mod resolves a PS4 ticket as a duplicate of an Xbox ticket with the KEEP_PLATFORM tag, that's either an incorrect resolution or the KEEP_PLATFORM tag was incorrectly set. In either case, a mod should look at the tickets and make an appropriate fix. Having a warning in the log will help to find these cases, though eventually sending a message directly would be better.

File: arisa.json (hunk 2 above)
Comment (chandler05) [reply arisa.json:197]: K, removed from the targetConfirmationStatus as well

**Comment by chandler05** [comment]
Or, the user could have put the wrong platform for the duplicate

**Review by urielsalis — APPROVED** [review]

File: src/main/kotlin/io/github/mojira/arisa/modules/MultiplePlatformsModule.kt (hunk 1 above)
Comment (SunCatMC) [reply src/main/kotlin/io/github/mojira/arisa/modules/MultiplePlatformsModule.kt:34]: getFullIssue() already returns an either, just a different one

and I don't see a cleaner way to deal with that honestly

and bind is needed 'cos when Either is Right it converts Either into its right half, that Boolean that you need

**Pull request merged**